File: src/main/java/com/fengsheng/skill/ShiSi.java
package com.fengsheng.skill;

import com.fengsheng.Game;
import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import com.fengsheng.phase.ReceivePhaseReceiverSkill;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Role;
import org.apache.log4j.Logger;

/**
 * 老汉技能【视死】：你接收黑色情报后，摸两张牌。
 */
public class ShiSi extends AbstractSkill implements TriggeredSkill {
    private static final Logger log = Logger.getLogger(ShiSi.class);

    @Override
    public SkillId getSkillId() {
        return SkillId.SHI_SI;
    }

    @Override
    public ResolveResult execute(Game g) {
        if (!(g.getFsm() instanceof ReceivePhaseReceiverSkill fsm) || fsm.inFrontOfWhom().findSkill(getSkillId()) == null || !fsm.inFrontOfWhom().isAlive())
            return null;
        if (fsm.inFrontOfWhom().getSkillUseCount(getSkillId()) > 0)
            return null;
        var colors = fsm.messageCard().getColors();
        if (!colors.contains(Common.color.Black))
            return null;
        fsm.inFrontOfWhom().addSkillUseCount(getSkillId());
        log.info(fsm.inFrontOfWhom() + "发动了[视死]");
        for (Player p : g.getPlayers()) {
            if (p instanceof HumanPlayer player)
                player.send(Role.skill_shi_si_toc.newBuilder().setPlayerId(player.getAlternativeLocation(fsm.inFrontOfWhom().location())).build());
        }
        fsm.inFrontOfWhom().draw(2);
        return null;
    }
}

File: src/main/kotlin/skill/ShiSi.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.protos.skillShiSiToc
import org.apache.logging.log4j.kotlin.logger

/**
 * 老汉技能【视死】：你接收黑色情报后，摸两张牌。
 */
class ShiSi : TriggeredSkill {
    override val skillId = SkillId.SHI_SI

    override val isInitialSkill = true

    override fun execute(g: Game, askWhom: Player): ResolveResult? {
        val event = g.findEvent<ReceiveCardEvent>(this) { event ->
            askWhom === event.inFrontOfWhom || return@findEvent false
            event.inFrontOfWhom.getSkillUseCount(skillId) == 0 || return@findEvent false
            event.messageCard.isBlack()
        } ?: return null
        logger.info("${event.inFrontOfWhom}发动了[视死]")
        g.players.send { skillShiSiToc { playerId = it.getAlternativeLocation(event.inFrontOfWhom.location) } }
        event.inFrontOfWhom.draw(2)
        return null
    }
}